src/env/log_fd/nix.rs

```diff
@@ -0,0 +1,206 @@
+// Copyright (c) 2017-present, PingCAP, Inc. Licensed under Apache-2.0.
+
+use crate::env::{Handle, Permission};
+
+use fail::fail_point;
+use log::error;
+
+use std::os::unix::io::RawFd;
+
+use nix::errno::Errno;
+use nix::fcntl::{self, OFlag};
+use nix::sys::stat::Mode;
+use nix::sys::uio::{pread, pwrite};
+use nix::unistd::{close, ftruncate, lseek, Whence};
+use nix::NixPath;
+
+fn from_nix_error(e: nix::Error, custom: &'static str) -> std::io::Error {
+    let kind = std::io::Error::from(e).kind();
+    std::io::Error::new(kind, custom)
+}
+
+impl From<Permission> for OFlag {
+    fn from(value: Permission) -> OFlag {
+        match value {
+            Permission::ReadOnly => OFlag::O_RDONLY,
+            Permission::ReadWrite => OFlag::O_RDWR,
+        }
+    }
+}
+
+/// A RAII-style low-level file. Errors occurred during automatic resource
+/// release are logged and ignored.
+///
+/// A [`LogFd`] is essentially a thin wrapper around [`RawFd`]. It's only
+/// supported on *Unix*, and primarily optimized for *Linux*.
+///
+/// All [`LogFd`] instances are opened with read and write permission.
+pub struct LogFd(RawFd);
+
+impl LogFd {
+    /// Opens a file with the given `path`.
+    pub fn open<P: ?Sized + NixPath>(path: &P, perm: Permission) -> IoResult<Self> {
+        fail_point!("log_fd::open::err", |_| {
```

Comment:
This is tricky. All the failpoints need to be ported as well otherwise some unit tests can fail.

Reply:
You mean the fail points should be enabled for both LogFd impl? WDYT if I made a new LogFdImpl trait, and use different impl in different env? Like this:

```rust
// in log_fd.rs:

trait LogFdImpl {
  fn foo(&self);
}

struct LogFd(Arc<dyn LogFdImpl>)

impl LogFd {
  fn foo(&self) {
    fail_points!(...)
    self.0.foo();
  }
}

// in log_fd/nix.rs:

struct NixLogFd

impl LogFdImpl for NixLogFd 

// in log_fd/plain.rs:

struct StdLogFd

impl LogFdImpl for StdLogFd

// then in log_fd.rs:

impl LogFd {
  fn open() -> Self {
    #[cfg(unix)]
    LogFd(Arc::new(logfd::nix::NixLogFd))
    #[cfg(not(unix))]
    LogFd(Arc::new(logfd::plain::StdLogFd))
  }
}
#[cfg(unix)]

```

Reply:
That seems too cumbersome. We can either port failpoints, or disable failpoint tests when std is used (using required-features):

https://github.com/tikv/raft-engine/blob/1c7a820b251e7d012f463e1aee722eac5b5a0440/Cargo.toml#L23-L26

Reply:
Oh, we can still move some of them to `LogFile` implementation. Only failpoints associated with methods from trait `Handle` is required to be implemented for `LogFd`.

Reply:
Not sure I've fully understood it, sorry, but do you mean: 1. the current `LogFd` implementations (called by same function signatures, enabled under different `cfg`s) are fine, don't need to be further changed? 2. I have to port the fail points in log_fd::plain, too?

Reply:
No,
1. (optionally) You can move some failpoints under `LogFd` to `LogFile`. Because `LogFd` only implements one public interface `Handle`, other methods are only called by `LogFile`. So moving those not associated with `Handle` will be not change the test behavior.
2. Duplicate the remaining failpoints under `LogFd` to `plain::LogFd`.